File: frontend/src/types/samples07.ts
/** Domain types for samples / SampleRecord07 */
export type SampleRecord07Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface SampleRecord07 {
  id: string;
  externalCode: string;
  displayName: string;
  status: SampleRecord07Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface SampleRecord07CreateRequest {
  externalCode: string;
  displayName: string;
  status: SampleRecord07Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface SampleRecord07ListResponse {
  items: SampleRecord07[];
  total?: number;
}

export interface SampleRecord07StatsResponse {
  activeCount: number;
}

export function isSampleRecord07Active(rec: SampleRecord07): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatSampleRecord07Label(rec: SampleRecord07): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareSampleRecord07ByPriority(a: SampleRecord07, b: SampleRecord07): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/samples_07.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// SampleRecord07Service encapsulates business rules for SampleRecord07.
type SampleRecord07Service struct {
	repo *repository.SampleRecord07Repository
}

// NewSampleRecord07Service wires the service.
func NewSampleRecord07Service(repo *repository.SampleRecord07Repository) *SampleRecord07Service {
	return &SampleRecord07Service{repo: repo}
}

// Create validates and persists a new SampleRecord07.
func (s *SampleRecord07Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.SampleRecord07, error) {
	now := time.Now().UTC()
	rec := &models.SampleRecord07{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a SampleRecord07 by id.
func (s *SampleRecord07Service) Get(ctx context.Context, id string) (*models.SampleRecord07, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped SampleRecord07 pages.
func (s *SampleRecord07Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.SampleRecord07, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing SampleRecord07.
func (s *SampleRecord07Service) Update(ctx context.Context, actor string, rec *models.SampleRecord07) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a SampleRecord07.
func (s *SampleRecord07Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds SampleRecord07 by external code prefix.
func (s *SampleRecord07Service) Search(ctx context.Context, facilityID, prefix string) ([]models.SampleRecord07, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *SampleRecord07Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
